File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/string/_205_Isomorphic_Strings.java
package io.hank.leetcode.practices.string;

import io.hank.leetcode.annotations.*;
import io.hank.leetcode.practices.LeetcodeProblemSolution;

import java.util.Arrays;

/**
 * <pre>
 * Given two strings s and t, determine if they are isomorphic.
 *
 * Two strings s and t are isomorphic if the characters in s can be replaced to get t.
 *
 * All occurrences of a character must be replaced with another character while preserving the order of characters. No two characters may map to the same character, but a character may map to itself.
 *
 *
 * Example 1:
 *
 * Input: s = "egg", t = "add"
 *
 * Output: true
 *
 * Explanation:
 *
 * The strings s and t can be made identical by:
 *
 * Mapping 'e' to 'a'.
 * Mapping 'g' to 'd'.
 *
 * Example 2:
 *
 * Input: s = "foo", t = "bar"
 *
 * Output: false
 *
 * Explanation:
 *
 * The strings s and t can not be made identical as 'o' needs to be mapped to both 'a' and 'r'.
 *
 * Example 3:
 *
 * Input: s = "paper", t = "title"
 *
 * Output: true
 *
 *
 * Constraints:
 *
 * 1 <= s.length <= 5 * 104
 * t.length == s.length
 * s and t consist of any valid ascii character.
 * </pre>
 */
public class _205_Isomorphic_Strings extends LeetcodeProblemSolution {

    @Topic(TopicType.STRING)
    @TimeComplexity(ComplexityType.O_N)
    @SpaceComplexity(ComplexityType.O_1)
    public boolean isIsomorphic(String s, String t) {
        int[] sToT = new int[256];
        int[] tToS = new int[256];
        Arrays.fill(sToT, -1);
        Arrays.fill(tToS, -1);
        for (int i = 0; i < s.length(); i++) {
            char sChar = s.charAt(i);
            char tChar = t.charAt(i);
            if (sToT[sChar] == -1 && tToS[tChar] == -1) {
                sToT[sChar] = tChar;
                tToS[tChar] = sChar;
            } else {
                if (sToT[sChar] != tChar || tToS[tChar] != sChar) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: leetcode-practice-typescript/src/hash/205-isomorphic-strings.ts
function isIsomorphic(s: string, t: string): boolean {
    if (s.length !== t.length) {
        return false;
    }
    if (s.length === 1) {
        return s === t;
    }

    const stMapping = new Map<string, string>();
    const tsMapping = new Map<string, string>();
    for (let i = 0; i < s.length; i++) {
        const sChar = s[i];
        const tChar = t[i];
        if (stMapping.has(sChar)) {
            if (tChar !== stMapping.get(sChar)) {
                return false;
            }
        } else if (tsMapping.has(tChar)) {
            // stMapping没有但tsMapping却有
            return false;
        } else {
            stMapping.set(sChar, tChar);
            tsMapping.set(tChar, sChar);
        }
    }
    return true;
}

console.log(isIsomorphic("badc", "baba"));
console.log(isIsomorphic("paper", "title"));
